New Article: Pitfalls of Safe Rust

mre opened the pull request.

mre committed:
Article "Safe Rust Isn’t Always Safe"

mre committed:
Update article structure and improve clarity

mre committed:
fix typos

mre committed:
date

mre committed:
CTA

mre committed:
order

mre committed:
TOC

mre committed:
cleanup

mre committed:
streamline intro

mre committed:
Redundant headline lol

wezm commented on content/blog/safe-rust-isnt-always-safe/index.md:
```diff
@@ -0,0 +1,692 @@
+{% info(title="Quick Tip: Enable Overflow Checks In Release Mode", icon="info") %}
+
+In general, Rust values performance at least as much as safety.
+Overflow checks are expensive, which is why Rust disables them in release mode.
```
> Overflow checks are expensive

Might be worth backing this up with some evidence. Here's an older post from Dan Luu https://danluu.com/integer-overflow/ which suggests the checking isn't too bad but it can have an impact on other optimisations.

mre replied:
Thanks! It's probably outside of the scope of this post, but I'll add the link for further reference and change the wording to "can be expensive."

wezm commented on content/blog/safe-rust-isnt-always-safe/index.md:
```diff
@@ -0,0 +1,692 @@
+
+impl std::fmt::Debug for Password {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("[REDACTED]")
```
One drawback of manually implementing `Debug` is that it can get out of sync with the type, which in the past led me to be a bit resistant to making custom impls. A pattern I've seen to address this is to destructure the value so you get an error if there are incorrect fields:

```
impl fmt::Debug for Settings {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let Settings {
            content_dir,
            author,
            fediverse_creator,
            copyright_start_year,
            name,
            description,
            sentry_dsn,
            base_url,
            timezone,
            version,
            render_future_posts,
        } = self;
        f.debug_struct("Settings")
            .field("content_dir", content_dir)
            .field("author", author)
            .field("fediverse_creator", fediverse_creator)
            .field("copyright_start_year", copyright_start_year)
            .field("name", name)
            .field("description", description)
            .field("sentry_dsn", sentry_dsn)
            .field("base_url", &base_url.as_str())
            .field("timezone", timezone)
            .field("version", version)
            .field("render_future_posts", render_future_posts)
            .finish()
    }
}
```

mre replied:
Wow, that's pretty nice! I'll add that to a "details" section.

wezm commented on content/blog/safe-rust-isnt-always-safe/index.md:
Not sure it's swallowing errors, since it's intended behaviour.

wezm reviewed (commented):
Another great post. Left a few comments that you may or may not choose to ignore :)

mre committed:
update post from user feedback

mre committed:
rename folder

mre committed:
Remove part about thread-safety

It was misleading and would need a broader context to be valuable. Could be a follow-up article

mre commented:
@wezm, thanks for all the great input. I've integrated it into the article and added you as a reviewer. 🚀

durka commented on content/blog/pitfalls-of-safe-rust/index.md:
````diff
@@ -0,0 +1,687 @@
+  = note: `#[deny(arithmetic_overflow)]` on by default
+```
+
+For all other cases, use [`checked_add`](https://docs.rs/num/latest/num/trait.CheckedAdd.html), [`checked_sub`](https://docs.rs/num/latest/num/trait.CheckedSub.html), [`checked_mul`](https://docs.rs/num/latest/num/trait.CheckedMul.html), and [`checked_div`](https://docs.rs/num/latest/num/trait.CheckedDiv.html) to avoid overflow.
````
You might want to also mention the wrapping and saturating methods for completeness

mre replied:
Great idea; added a note!

durka commented on content/blog/pitfalls-of-safe-rust/index.md:
```diff
@@ -0,0 +1,687 @@
+
+{% info(title="Quick Tip: Enable Overflow Checks In Release Mode", icon="info") %}
+
+In general, Rust values performance at least as much as safety.
```
Not sure I agree with this statement. 

Rust made the decision that integer overflow isn't unsafe. But in other scenarios a performance hit is acceptable to ensure memory safety, if there's no alternative. 

One example would be checked array indexing. Another is when they fixed float casting: https://internals.rust-lang.org/t/help-us-benchmark-saturating-float-casts/6231

mre replied:
Great idea. Added a note!

mre replied:
Thanks. I've updated the wording. Can you check again to see if it's okay now?

durka commented on content/blog/pitfalls-of-safe-rust/index.md:
````diff
@@ -0,0 +1,687 @@
+        Ok(UserCredentials { username, password })
+    }
+}
+```
````
I'm questioning if manually implementing that traits is the best recommendation. Forgoing the derive seems error-prone. 

Furthermore, this issue isn't just with deserialization! I could just create an instance of the struct in code with empty strings and that would be invalid, so it kind of goes back to the earlier point about using types to make invalid things unrepresentable. In that vein, I found this article, which proposes a validation strategy for deserialization. So the idea would be to use a new type like NonemptyString for the fields. https://dev.to/equalma/validate-fields-and-types-in-serde-with-tryfrom-c2n

mre replied:
Oh, I didn't know about that approach! This is amazing. I will replace my recommendation with the suggestion from the article you linked to.

mre replied:
Added. Can you check again?

durka commented on content/blog/pitfalls-of-safe-rust/index.md:
I don't know much about this area but why check the length first? Isn't that another timing leak that could be used to deduce the length?

mre replied:
Ah, gotcha. My performance optimization backfired! Let me remove that again. Thanks for pointing that out.

durka reviewed (commented):
Done reviewing! I think this is a great article! 

My main quibble is with the definition of "safe". Personally I always want it to mean "memory safe" when referring to Rust. "Safe Rust isn't always safe" is just a tiny bit clickbaity, because it sounds like you're going to talk about violating memory safety in safe code, which of course is possibly via soundness bugs or that /proc/mem trick but that's not what your post is about. 

I guess I might go with something more like "memory safety isn't the only kind of safety" but I don't want to prescribe you a title! 

Overall, 👍 👍

mre commented:
@durka, thanks so much for the review! 

I changed the title now. Hopefully it's less ambiguous. 😅

mre committed:
update

mre committed:
update header section

mre committed:
Improve content on serialization best practices.

mre committed:
Remove insecure perf optimization

mre committed:
Clarifications and fix typos

mre commented:
@durka, I got around to integrating the rest of your feedback. I'm really happy with how it turned out, thanks! If you find the time, please double-check. If not, that's also fine. I've added you as a reviewer. Thanks a lot! 😊

durka commented:
Looks good to me! :)

mre commented:
Thanks for the feedback!

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
Mhh. Not sure I would call it "safety", maybe "data integrity"? The problem I have is that a panic at this point can be unsafe too. And it's _very_ hidden. Especially in server code I wouldn't call it "safe".

mre replied:
Good point. How about simply naming what the feature actually does, better overflow detection?

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
```suggestion
- use `From::from()` when you can guarantee no data loss.
```

If I wanna turn my `i64` into a `u8` and "need to guarantee no data loss" I still can't use `From::from()`. When I "need to guarantee no data loss" on `i64` to `u8` then I have to use `TryFrom` and handle the error case well.

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
```suggestion
- only use `as` when you're comfortable with potential truncation or know the values will fit within the target type's range and when performance is absolutely critical.
```
Like this? Seems to be somewhat implied by the ⚠️ sign above, but not clearly stated until after this line.

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
```suggestion
let y = i8::try_from(x).ok_or("Number is too big to be used here")?;
```
More concise and real-world related.

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
Why should I avoid this? It's not clear to me from just reading this section. I assume you mean this should be encapsulated so we don't have invalid UserIDs by just coming up with numbers, but instead having to use an API that only hands our valid UserIDs which we could then use in other places of that API?

This reads like a very slightly different advice than the previous section, but the difference isn't clearly stated.

mre replied:
I've updated the example. Can you check again?

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
This is a bad example in this section. You were just talking about redacting fields, then about potentially forgetting to update the impl on updates to the struct, but then in this example you could just use the derived Debug again.

A better example could be a database connection URI? You usually need passwords for DB access, and you don't want to accidentally leak them.

```rust
struct DatabaseURI {
    scheme: String,
    user: String,
    password: String,
    host: SocketAddr,
    database: String,
}

// don't
impl std::fmt::Debug for DatabaseURI {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}://{}:[REDACTED]@{}/{}", self.scheme, self.user, self.host, self.database)
    }
}

// do
impl std::fmt::Debug for DatabaseURI {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let DatabaseURI = { scheme, user, password: _, host, database } = self;
        write!(f, "{scheme}://{user}:[REDACTED]@{host}/{database}")
        // -- or --
        f.debug_struct("DatabaseURI")  
            .field("scheme", scheme)
            .field("user", user)
            .field("password", &"***")
            .field("host", host)
            .field("database", database)
            .finish()
    }
}
```

mre replied:
Good point!

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
"Don't blindly" could be remembered as "never". The "careful" from the heading is more appropriate.

mre replied:
I think the wording is clear? 🤔 You could interpret it this way, but it's not implied in the wording.

M3t0r replied:
I'm just saying that you want them to take away that they should be careful, and they could understand to never derive those traits.

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
These examples are so wildly different it's almost comical. In the first you want to read a file that has to exist, in the second you open a file to write regardless of if it exists or not?

And calling them "atomic file operations" is a bit wild to me too. They are not "arbitrary file operations", they are just a few options of [`open()`](https://www.man7.org/linux/man-pages/man2/open.2.html). Every unix hacker / C dev will know them.

The first example is a classic case of "just open the damn file, the OS will tell you if it doesn't exist".

```rust
// do
fn process_file(path: &Path) -> Result<(), io::Error> {
    let content = fs::read(path)?;  // read will fail if the path doesn't exist
    // Process content
    Ok(())
}
```

A _slightly_ better example:
```rust
let config = if config_is_valid(&path) {
    // Attacker could change file contents here!
    load_config(&path)
} else {
    return Err("Config is not valid")
}
```

mre replied:
I've updated it. Can you double-check?

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
```suggestion
const MAX_REQUEST_SIZE: usize = 1024 * 1024;  // 1MiB
```
1MB = 1000 * 1000 :wink:

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
That's not added safety, that's highly confusing and bad advice. What does the decode do? Does it strictly shorten the data? Can it increase the size? Or is it simply a 1:1 translation?

Just leave it out, but at the very least use a different constant.

mre replied:
Removed that part, thanks.

M3t0r commented on content/blog/pitfalls-of-safe-rust/index.md:
Are crypto crates "non-sys level"? Not quite sure this is good advice. Or even valid in this blanket statement. Maybe ask readers to investigate the use of `unsafe` in that crate and judge if it's necessary or harmful for their application?

mre replied:
True. Removed that sentence.

mre committed:
Update content/blog/pitfalls-of-safe-rust/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Update content/blog/pitfalls-of-safe-rust/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Update content/blog/pitfalls-of-safe-rust/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Update content/blog/pitfalls-of-safe-rust/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
User "overflow detection" as better expression

mre committed:
Explain difference to newtype pattern in field-combination example

mre committed:
Extend derive(Debug) example

mre committed:
Update example

mre committed:
Update TOCTOU example

mre committed:
Remove decoded size check

mre committed:
Clarify guidance on evaluating dependencies with unsafe code

mre committed:
update date

mre commented:
@M3t0r, thanks for the great advice. I've finally updated the post. Could you take another look?

mre committed:
Add link to std

mre committed:
Add "High Assurance Rust"

mre committed:
fix toml frontmatter

M3t0r reviewed (approved):
Looks good. Send it :+1:

mre committed:
Update date

1st of April, lol
